src/nifs/protogalaxy/poly/folded_trace.rs

```diff
@@ -0,0 +1,180 @@
+///
+/// Since the number of rows is large, we do this in one pass, counting the points for each
+/// challenge at each iteration, and laying them out in separate [`PlonkWitness`] at the end.
+fn fold_witnesses<F: PrimeField>(
+    X_challenges: &[F],
+    accumulator: &(impl GetWitness<F> + Sync),
+    witnesses: &[impl Sync + GetWitness<F>],
+) -> Vec<PlonkWitness<F>> {
+    let log_n = (witnesses.len() + 1).next_power_of_two().ilog2();
+
+    let lagrange_poly_by_challenge = X_challenges
+        .iter()
+        .map(|X| {
+            lagrange::iter_eval_lagrange_polynomials_for_cyclic_group(*X, log_n)
+                .collect::<Box<[_]>>()
+        })
+        .collect::<Box<[_]>>();
+
+    let empty_witness = accumulator
+        .get_witness()
+        .iter()
+        .map(|row| vec![F::ZERO; row.len()]) // TODO Use memory allocation with zeroes, for not iterate over `row.len()`
+        .collect::<Vec<_>>();
+
+    // TODO Create on the fly to avoid multiple rows iterations
+    let mut result_matrix_by_challenge = vec![
+        PlonkWitness {
+            W: empty_witness.clone(),
+        };
+        X_challenges.len()
+    ];
+
+    accumulator
+        .get_witness()
+        .iter()
+        .enumerate()
+        .flat_map(|(column, witness)| {
+            let rows_count = witness.len();
+            iter::repeat(column).zip(0..rows_count)
+        })
+        .map(|(col, row)| {
+            iter::once(accumulator.get_witness())
+                .chain(witnesses.iter().map(GetWitness::get_witness))
+                .zip(
+                    lagrange_poly_by_challenge
+                        .iter()
+                        .map(|m| m.iter().copied())
+                        .multi_product(),
+                )
+                .fold(
+                    vec![F::ZERO; X_challenges.len()].into_boxed_slice(),
+                    // every element of this collection - one cell for each `X_challenge`
+                    |mut cells_by_challenge, (witness, multiplier)| {
+                        cells_by_challenge
+                            .iter_mut()
+                            .zip(multiplier.iter())
+                            .for_each(|(res, cell)| {
+                                *res += *cell * witness[col][row];
+                            });
+
+                        cells_by_challenge
+                    },
+                )
+        })
+        .zip(
+            // Here we take an iterator that on each iteration returns [column][row] elements for
+            // each witness for its challenge
+            //
+            // next -> vec![result[0][col][row], result[1][col][row], ... result[challenges_len][col][row]]
+            result_matrix_by_challenge
+                .iter_mut()
+                .map(|matrix| matrix.W.iter_mut().flat_map(|col| col.iter_mut()))
+                .multi_product(),
+        )
+        .par_bridge()
+        .for_each(|(elements, mut results)| {
```

Comment:
what is the purpose of this last "for_each"?

Reply:
Put the data into pre-allocated memory, and given `par_bridge` - make this process paralleled